File: 42-trapping-rain-water/trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int left=0, right=height.length-1;
        int lmax=0,rmax=0;
        int watTapped = 0;

        while(left < right){
            if(height[left] < height[right]){
                if(height[left]>=lmax){
                    lmax=height[left];
                }else{
                    watTapped = watTapped + lmax - height[left];
                }
                left++;
            }else{
                if(height[right] >= rmax){
                    rmax=height[right];
                }else{
                    watTapped = watTapped + rmax - height[right];
                }
                right--;
            }
        }
        return watTapped;
        
    }
}

File: 42-trapping-rain-water/trapping-rain-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {

        let n = height.length;
        let left = 0;
        let right = n-1;

        let left_max=height[left],right_max=height[right];
        let water = 0;

        while(left < right){
            if(left_max < right_max){
                left++;
                left_max = Math.max(left_max,height[left]);
                water += left_max - height[left];
            }else{
                right--;
                right_max = Math.max(right_max,height[right]);
                water += right_max - height[right];
            }
        }
        return water;
    
};